### PR TITLE
Learn conditional factoring rules in rational

### PR DESCRIPTION
**Pull request by oflatt**
This PR significantly improves the rational suite by adding a recipe that learns factoring rules.
It also changes the constant folder from `Ratio<BigInteger>` to `Rational64` for performance.

### CODE REVIEW

File: src/enumo/ruleset.rs [hunk 1]
```diff
@@ -432,18 +436,28 @@ impl<L: SynthLanguage> Ruleset<L> {
...
+            print!("Selected: ");
+            selected.pretty_print();
             chosen.extend(selected.clone());
             self.shrink(&chosen, scheduler);
         }
         // Return only the new rules
         chosen.remove_all(prior);
 
-        chosen
+        (chosen, invalid)
```
Comment (oflatt) [inline src/enumo/ruleset.rs:460]: return the invalid rules so we can learn things conditionally

File: src/recipe_utils.rs [hunk 2]
```diff
@@ -16,6 +16,15 @@ pub fn iter_metric(wkld: Workload, atom: &str, met: Metric, n: usize) -> Workloa
     pegs
 }
 
+pub fn substitute(wklod: Workload, sub: Workload, atom: &str) -> Workload {
```
Comment (oflatt) [inline src/recipe_utils.rs:19]: useful helper, but unused right now- i could remove

File: tests/rational.rs [hunk 3]
```diff
@@ -123,35 +159,57 @@ impl SynthLanguage for Math {
...
-        }
-
-        let rhs_errors = Self::one_of_errors(&ctx, rhs_denom);
-        let lhs_errors = Self::one_of_errors(&ctx, lhs_denom);
-        let error_preserved = rhs_errors.iff(&lhs_errors);
-        let assertion = z3::ast::Bool::and(&ctx, &[&assert_equal, &error_preserved]);
-
-        solver.assert(&assertion.not());
-        Self::z3_res_to_validationresult(solver.check())
+        solver.assert(&assert_equal.not());
```
Comment (oflatt) [inline tests/rational.rs:206]: greatly simplified- now we define division by zero to be 1

File: tests/rational.rs [hunk 4]
```diff
@@ -349,10 +455,13 @@ fn egg_to_z3<'a>(ctx: &'a z3::Context, expr: &[Math]) -> z3::ast::Real<'a> {
...
             )),
             Math::Div([x, y]) => {
-                // Do NOT assume denominator is non-zero
-                buf.push(z3::ast::Real::div(
-                    &buf[usize::from(*x)],
-                    &buf[usize::from(*y)],
+                let zero = z3::ast::Real::from_real(ctx, 0, 1);
+                let one = z3::ast::Real::from_real(ctx, 1, 1);
+                // Division by zero set to one
+                buf.push(z3::ast::Bool::ite(
```
Comment (oflatt) [inline tests/rational.rs:461]: Important

File: tests/rational.rs [hunk 5]
```diff
@@ -695,40 +795,6 @@ pub mod test {
         );
     }
 
-    #[test]
```
Comment (oflatt) [inline tests/rational.rs:698]: no longer works with new semantics

File: src/enumo/ruleset.rs [hunk 6]
```diff
@@ -432,18 +436,28 @@ impl<L: SynthLanguage> Ruleset<L> {
...
         while !self.is_empty() {
-            let selected = self.select(step_size);
+            println!(
+                "Shrinking {}/{} candidates. Kept {} so far.",
+                self.len(),
+                before,
+                chosen.len()
+            );
+            let selected = self.select(step_size, &mut invalid);
+            print!("Selected: ");
```
Comment (ajpal) [inline src/enumo/ruleset.rs:452]: nit: remove printlns if you don't need them anymore
(fine to do as followup if you want to get this merged soon!)

File: src/recipe_utils.rs [hunk 7]
```diff
@@ -34,7 +43,7 @@ pub fn run_workload<L: SynthLanguage>(
     };
 
     let num_prior = prior.len();
-    let chosen = candidates.minimize(prior, Scheduler::Compress(limits));
+    let chosen = candidates.minimize(prior, Scheduler::Compress(limits)).0;
```
Comment (ajpal) [inline src/recipe_utils.rs:46]: style nit: prefer `let (chosen, _) = ...` over `let chosen = <something>.0`

File: tests/rational.rs [hunk 8]
```diff
@@ -311,12 +332,16 @@ impl Math {
             "Instrumented {} rules with conditions",
             with_condition.len()
         );
+        /*println!("Conditioned rules:");
```
Comment (ajpal) [inline tests/rational.rs:335]: remove

**Pull request merged**